File: BinarySearch/C++/875KokoEatingBananas.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        // Create search range
        int left = 1;
        int right = 0;
        // Get the max element in the vector, kind of shitty in C++ , its equivalent of max()
        // in python
        for (int i : piles) {
            right = max(right, i);
        }
        int result = right;
  
        // Start binary serach
        while (left <= right){
            int k = ((right - left) / 2) + left;
            
            // Check if we all bananas can be eaten with this rate
            long int hours = 0;
            for (int i: piles){
                hours += std::ceil((double) i / k);
            }
            if (hours <= h){
                right = k - 1;
                result = std::min(result, k);

            } else {
                left = k + 1; 
            }
        }

        return result;
        
        
    }
};

File: BinarySearch/Python/875KokoEatingBananas.py
# Time Complexity: O(PLogK)
# Space Complexity: O(1)
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
    	# Define a search range
    	l = 1
    	r = max(piles)
    	result = r

    	# Use binary search to find the lowest eating rate
    	while l <= r:
    		k = ((r - l) // 2) + l
    		hours = 0
    		# Hours taken to eat all the bananas given the current eating rate
    		for p in piles:
    			hours += math.ceil( p / k)
    		if hours <= h:
    			result = k
    			r = k - 1
    		else:
    			l = k + 1
    	return result


